File: client-norx/app/src/main/java/ua/in/meugen/myapplication/ui/activities/main/fragment/MainFragmentModule.java
package ua.in.meugen.myapplication.ui.activities.main.fragment;

import android.arch.lifecycle.ViewModelProvider;
import android.arch.lifecycle.ViewModelProviders;
import android.support.v4.app.Fragment;

import dagger.Binds;
import dagger.Module;
import dagger.Provides;
import ua.in.meugen.myapplication.app.di.scopes.PerFragment;
import ua.in.meugen.myapplication.model.actions.AppActionApi;
import ua.in.meugen.myapplication.model.actions.items.ItemsActionApi;
import ua.in.meugen.myapplication.model.actions.items.ItemsRequest;
import ua.in.meugen.myapplication.model.network.Resource;
import ua.in.meugen.myapplication.model.network.resp.ItemsResponse;
import ua.in.meugen.myapplication.ui.activities.base.fragment.BaseFragment;
import ua.in.meugen.myapplication.ui.activities.base.fragment.BaseFragmentModule;
import ua.in.meugen.myapplication.ui.activities.main.fragment.vm.MainViewModel;

@Module(includes = BaseFragmentModule.class)
public abstract class MainFragmentModule {

    @Binds @PerFragment
    abstract BaseFragment bindBaseFragment(final MainFragment fragment);

    @Provides @PerFragment
    static MainViewModel provideMainViewModel(
            final Fragment fragment,
            final ViewModelProvider.Factory factory) {
        return ViewModelProviders.of(fragment, factory).get(MainViewModel.class);
    }
}

File: client-coroutines/app/src/main/java/ua/meugen/android/client/coroutines/ui/activities/main/fragment/MainFragmentModule.kt
package ua.meugen.android.client.coroutines.ui.activities.main.fragment

import android.support.v4.app.Fragment
import dagger.Binds
import dagger.Module
import ua.meugen.android.client.coroutines.app.di.PerFragment
import ua.meugen.android.client.coroutines.ui.activities.base.fragment.BaseFragmentModule
import ua.meugen.android.client.coroutines.ui.activities.base.fragment.state.EmptyState
import ua.meugen.android.client.coroutines.ui.activities.base.fragment.state.MvpState
import ua.meugen.android.client.coroutines.ui.activities.main.fragment.presenter.MainPresenter
import ua.meugen.android.client.coroutines.ui.activities.main.fragment.presenter.MainPresenterImpl
import ua.meugen.android.client.coroutines.ui.activities.main.fragment.view.MainView

@Module(includes = [BaseFragmentModule::class])
abstract class MainFragmentModule {

    @Binds @PerFragment
    abstract fun bindFragment(fragment: MainFragment): Fragment

    @Binds @PerFragment
    abstract fun bindPresenter(presenter: MainPresenterImpl): MainPresenter

    @Binds @PerFragment
    abstract fun bindState(state: EmptyState): MvpState

    @Binds @PerFragment
    abstract fun bindView(fragment: MainFragment): MainView
}